core/embed/trezorhal/unix/touch/touch.c

```diff
@@ -31,17 +31,145 @@
 extern int sdl_display_res_x, sdl_display_res_y;
 extern int sdl_touch_offset_x, sdl_touch_offset_y;
 
-static bool _touch_detected = false;
 static int _touch_x = 0;
 static int _touch_y = 0;
 
+// distance from the edge where arrow button swipe starts [px]
+static const int _btn_swipe_begin = 20;
+// length of the arrow button swipe [px]
+static const int _btn_swipe_length = 60;
+
+// A state machine to handle both mouse inputs (simulating touch) and arrow
+// buttons (to simulate scroll movements). The variable `input_state` is used to
+// ensure that arrow keys are not processed when mouse input is in progress and
+// that mouse actions are not processed while arrow button swipe is not
+// finished.
+typedef enum {
+  IDLE,
+  MOUSE_ACTIVE,
+  MOUSE_OUTSIDE,
+  BUTTON_SWIPE_INITIATED,
+  BUTTON_SWIPE_COMPLETED
+} TouchInputState;
+
+static TouchInputState input_state = IDLE;
+
 static bool is_inside_display(int x, int y) {
-  return (x >= sdl_touch_offset_x && y >= sdl_touch_offset_y &&
-          x - sdl_touch_offset_x < sdl_display_res_x &&
-          y - sdl_touch_offset_y < sdl_display_res_y);
+  return x >= sdl_touch_offset_x && y >= sdl_touch_offset_y &&
+         x - sdl_touch_offset_x < sdl_display_res_x &&
+         y - sdl_touch_offset_y < sdl_display_res_y;
+}
+
+void handle_mouse_events(SDL_Event event, int* ev_type, int* ev_x, int* ev_y) {
```

Comment:
All private function should be defined as `static`.

Reply:
refactored: https://github.com/trezor/trezor-firmware/pull/3619/commits/f248d6218433721716ffe2372555eb65fdeb0fc2